### PR TITLE
Automatically generate dune rules for tests

### PR DESCRIPTION
**Pull request by trefis**
Inspired by dune's own setup.

### CODE REVIEW

**Comment by trefis** [comment]
The error is due to me not generating a `(package merlin)` field in the rule stanza. I'll fix that tomorrow and then it should be good to merge.

File: tests/dune [hunk 1]
```diff
@@ -11,3 +11,14 @@
...
+(include dune.inc)
+
+(rule
+ (targets dune.inc.gen)
+ (deps    (source_tree .))
+ (action  (with-stdout-to %{targets} (run ./gen/gen_tests.exe))))
+
+(alias
+ (name   runtest)
+ (action (diff dune.inc dune.inc.gen)))
```
Comment (rgrinberg) [inline tests/dune]: Ddi you try `(mode promote)` in the `rule` above and change the target to `dune.inc`? That seems a bit simpler.

Comment (trefis) [reply tests/dune]: I just copied what dune itself is doing.
I think there was a reason why simply generating `dune.inc` wouldn't work, but I don't remember. Perhaps @aalekseyev will know?

**Comment by trefis** [comment]
Note: I ended up not generating the `(package merlin)` field, since LSP moved out.

**Comment by trefis** [comment]
Here is something I should have seen coming: just reading the tests directories to generate the rule means that the order in which rules are generated is not deterministic… which means travis is unhappy.

**Comment by rgrinberg** [comment]
> Here is something I should have seen coming: just reading the tests directories to generate the rule means that the order in which rules are generated is not deterministic… which means travis is unhappy.

In dune, I think we sort the dirs first.

**Pull request merged**